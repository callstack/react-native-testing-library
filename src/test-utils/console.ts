import { format } from 'util';

export function excludeConsoleMessage(logFn: (...args: unknown[]) => void, excludeMessage: string) {
  return (...args: unknown[]) => {
    const message = format(...args);
    if (message.includes(excludeMessage)) {
      return;
    }

    logFn(...args);
  };
}
