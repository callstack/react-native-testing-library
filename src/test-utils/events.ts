interface EventEntry {
  name: string;
  payload: any;
}

export function createEventLogger() {
  const events: EventEntry[] = [];
  const logEvent = (name: string) => {
    return (event: unknown) => {
      const eventEntry: EventEntry = {
        name,
        payload: event,
      };

      events.push(eventEntry);
    };
  };

  return { events, logEvent };
}
