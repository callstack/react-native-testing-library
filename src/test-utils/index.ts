export * from './console';
export * from './events';
