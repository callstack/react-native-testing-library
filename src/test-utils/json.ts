import type { ReactTestRendererJSON } from 'react-test-renderer';

type JsonPropsMapper = {
  [key: string]: unknown;
};

export function mapJsonProps<T extends ReactTestRendererJSON | ReactTestRendererJSON[] | null>(
  element: T,
  mapper: JsonPropsMapper,
): T {
  if (Array.isArray(element)) {
    return element.map((e) => mapJsonProps(e, mapper)) as T;
  }

  if (!element) {
    return element;
  }

  const resultProps = { ...element.props };
  Object.keys(mapper).forEach((key) => {
    if (key in element.props) {
      resultProps[key] = mapper[key];
    }
  });

  const resultElement = { ...element, props: resultProps };
  Object.defineProperty(resultElement, '$$typeof', {
    value: Symbol.for('react.test.json'),
  });

  return resultElement;
}
