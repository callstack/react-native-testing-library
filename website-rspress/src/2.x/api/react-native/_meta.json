[
  {
    "type": "file",
    "name": "index",
    "label": "Table of Contents"
  },
  {
    "type": "dir",
    "name": "classes",
    "label": "Classes"
  },
  {
    "type": "dir",
    "name": "interfaces",
    "label": "Interfaces"
  },
  {
    "type": "dir",
    "name": "types",
    "label": "Types"
  }
]
