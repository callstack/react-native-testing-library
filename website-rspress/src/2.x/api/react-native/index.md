# Re.Pack React Native APIs

## Table of contents

### Classes

- [Chunk](./classes/Chunk.md)
- [ChunkManager](./classes/ChunkManager.md)

### Interfaces

- [ChunkConfig](./interfaces/ChunkConfig.md)
- [ChunkManagerConfig](./interfaces/ChunkManagerConfig.md)
- [RemoteChunkLocation](./interfaces/RemoteChunkLocation.md)
- [StorageApi](./interfaces/StorageApi.md)

### Type aliases

- [RemoteChunkResolver](./types/RemoteChunkResolver.md)
