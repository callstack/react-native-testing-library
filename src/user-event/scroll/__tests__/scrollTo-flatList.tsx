import * as React from 'react';
import { FlatList, ScrollViewProps, Text } from 'react-native';
import { EventEntry, createEventLogger } from '../../../test-utils';
import { render, screen } from '../../..';
import { userEvent } from '../..';

const data = ['A', 'B', 'C', 'D', 'E', 'F', 'G'];

function mapEventsToShortForm(events: EventEntry[]) {
  return events.map((e) => [
    e.name,
    e.payload.nativeEvent.contentOffset.y,
    e.payload.nativeEvent.contentOffset.x,
  ]);
}

function renderFlatListWithToolkit(props: ScrollViewProps = {}) {
  const { events, logEvent } = createEventLogger();

  const renderItem = (title: string) => <Text>{title}</Text>;
  render(
    <FlatList
      testID="flatList"
      onScroll={logEvent('scroll')}
      onScrollBeginDrag={logEvent('scrollBeginDrag')}
      onScrollEndDrag={logEvent('scrollEndDrag')}
      onMomentumScrollBegin={logEvent('momentumScrollBegin')}
      onMomentumScrollEnd={logEvent('momentumScrollEnd')}
      onScrollToTop={logEvent('scrollToTop')}
      onEndReached={logEvent('endReached')}
      onViewableItemsChanged={logEvent('viewableItemsChanged')}
      data={data}
      renderItem={({ item }) => renderItem(item)}
      {...props}
    />
  );

  return { events };
}

describe('scrollTo() with FlatList', () => {
  it('supports vertical drag scroll', async () => {
    const { events } = renderFlatListWithToolkit();
    const user = userEvent.setup();

    await user.scrollTo(screen.getByTestId('flatList'), { y: 100 });
    expect(mapEventsToShortForm(events)).toEqual([
      ['scrollBeginDrag', 0, 0],
      ['scroll', 25, 0],
      ['scroll', 50, 0],
      ['scroll', 75, 0],
      ['scrollEndDrag', 100, 0],
    ]);
    expect(events).toMatchSnapshot('scrollTo({ y: 100 })');
  });

  it('supports horizontal drag scroll', async () => {
    const { events } = renderFlatListWithToolkit({ horizontal: true });
    const user = userEvent.setup();

    await user.scrollTo(screen.getByTestId('flatList'), { x: 100 });
    expect(mapEventsToShortForm(events)).toEqual([
      ['scrollBeginDrag', 0, 0],
      ['scroll', 0, 25],
      ['scroll', 0, 50],
      ['scroll', 0, 75],
      ['scrollEndDrag', 0, 100],
    ]);
  });
});
