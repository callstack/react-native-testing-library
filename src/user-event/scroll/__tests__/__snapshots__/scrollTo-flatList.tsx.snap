// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrollTo() with FlatList supports vertical drag scroll: scrollTo({ y: 100 }) 1`] = `
[
  {
    "name": "scrollBeginDrag",
    "payload": {
      "nativeEvent": {
        "contentInset": {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "contentOffset": {
          "x": 0,
          "y": 0,
        },
        "contentSize": {
          "height": 0,
          "width": 0,
        },
        "layoutMeasurement": {
          "height": 0,
          "width": 0,
        },
        "responderIgnoreScroll": true,
        "target": 0,
        "velocity": {
          "x": 0,
          "y": 0,
        },
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "nativeEvent": {
        "contentInset": {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "contentOffset": {
          "x": 0,
          "y": 25,
        },
        "contentSize": {
          "height": 0,
          "width": 0,
        },
        "layoutMeasurement": {
          "height": 0,
          "width": 0,
        },
        "responderIgnoreScroll": true,
        "target": 0,
        "velocity": {
          "x": 0,
          "y": 0,
        },
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "nativeEvent": {
        "contentInset": {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "contentOffset": {
          "x": 0,
          "y": 50,
        },
        "contentSize": {
          "height": 0,
          "width": 0,
        },
        "layoutMeasurement": {
          "height": 0,
          "width": 0,
        },
        "responderIgnoreScroll": true,
        "target": 0,
        "velocity": {
          "x": 0,
          "y": 0,
        },
      },
    },
  },
  {
    "name": "scroll",
    "payload": {
      "nativeEvent": {
        "contentInset": {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "contentOffset": {
          "x": 0,
          "y": 75,
        },
        "contentSize": {
          "height": 0,
          "width": 0,
        },
        "layoutMeasurement": {
          "height": 0,
          "width": 0,
        },
        "responderIgnoreScroll": true,
        "target": 0,
        "velocity": {
          "x": 0,
          "y": 0,
        },
      },
    },
  },
  {
    "name": "scrollEndDrag",
    "payload": {
      "nativeEvent": {
        "contentInset": {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "contentOffset": {
          "x": 0,
          "y": 100,
        },
        "contentSize": {
          "height": 0,
          "width": 0,
        },
        "layoutMeasurement": {
          "height": 0,
          "width": 0,
        },
        "responderIgnoreScroll": true,
        "target": 0,
        "velocity": {
          "x": 0,
          "y": 0,
        },
      },
    },
  },
]
`;
